### PR TITLE
Add bazel/cc_proto_descriptor_library

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
In preparation for supporting lite protos, add cc_proto_descriptor_library which can be used to provide 
proto descriptor capabilities even with lite protos. Based on previously closed-source library.

Risk Level: Low - Unused
Testing: New unit test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
/assign @abeyad 
Ok, looks like this is basically good to to!

File: bazel/cc_proto_descriptor_library/file_descriptor_generator.cc [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  contents << R"text(#include "bazel/cc_proto_descriptor_library/file_descriptor_info.h"
+)text";
+
+  google::protobuf::FileDescriptorProto file_descriptor_proto;
+  file->CopyTo(&file_descriptor_proto);
+  // XXX
+  /*
+    google::protobuf::FileDescriptorProto file_descriptor_proto =
+      google::protobuf::compiler::StripSourceRetentionOptions(*file);
+  */
```
Comment (abeyad) [inline bazel/cc_proto_descriptor_library/file_descriptor_generator.cc:115]: delete?

File: bazel/cc_proto_descriptor_library/file_descriptor_generator.cc [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+  contents << absl::StrFormat("#include \"%s\"\n", getDescriptorHeaderName(file));
+  for (int i = 0; i < file->dependency_count(); ++i) {
+    contents << absl::StrFormat("#include \"%s\"\n", getDescriptorHeaderName(file->dependency(i)));
+  }
+  contents << R"text(#include "bazel/cc_proto_descriptor_library/file_descriptor_info.h"
+)text";
+
+  google::protobuf::FileDescriptorProto file_descriptor_proto;
+  file->CopyTo(&file_descriptor_proto);
```
Comment (abeyad) [inline bazel/cc_proto_descriptor_library/file_descriptor_generator.cc:110]: in google3, this seems to call `proto2::compiler::StripSourceRetentionOptions` instead, not sure I understand the difference in what's happening here?  but if it works, i trust it :)

File: bazel/cc_proto_descriptor_library/repro/repro_test.cc [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+// Author: kmensah@google.com (Kwasi Mensah)
```
Comment (abeyad) [inline bazel/cc_proto_descriptor_library/repro/repro_test.cc:2]: not sure if we want to include this here?

File: bazel/cc_proto_descriptor_library/repro/repro_test.cc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+#include <string>
+
+#include "absl/strings/substitute.h"
+#include "bazel/cc_proto_descriptor_library/repro/base_with_any.pb.h"
+#include "bazel/cc_proto_descriptor_library/repro/base_with_any_descriptor.pb.h"
+#include "bazel/cc_proto_descriptor_library/text_format_transcoder.h"
+
```
Comment (abeyad) [inline bazel/cc_proto_descriptor_library/repro/repro_test.cc:10]: i don't see any test targets, could you please elaborate on what the test is intended to do?  (i'm also not clear on what the name `repro` means here?)

**Review by abeyad — COMMENTED** [review]
This is amazing!  Just left a few minor comments/questions.

File: api/bazel/api_build_system.bzl [hunk 5]
```diff
@@ -128,6 +133,19 @@ def api_cc_py_proto_library(
         deps = deps + _COMMON_PROTO_DEPS,
         visibility = visibility,
     )
+
+    # This is to support Envoy Mobile using Protobuf-Lite in google3.
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply api/bazel/api_build_system.bzl:137]: Good point. Removed.

File: api/bazel/api_build_system.bzl [hunk 6]
```diff
@@ -128,6 +133,19 @@ def api_cc_py_proto_library(
...
+    # Protobuf-Lite generated C++ code does not include reflection
+    # capabilities but analogous functionality can be provided by
+    # cc_proto_descriptor_library.
+    #
+    # go/envoy-mobile-protos
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply api/bazel/api_build_system.bzl:142]: Also a good point! :)

File: bazel/cc_proto_descriptor_library/file_descriptor_generator.cc (hunk 2 above)
Comment (RyanTheOptimist) [reply bazel/cc_proto_descriptor_library/file_descriptor_generator.cc]: Oh, right. That's in the code commented out below. Now that we're using 23.1 I can use that. Thanks for the reminder!

File: bazel/cc_proto_descriptor_library/file_descriptor_generator.cc (hunk 1 above)
Comment (RyanTheOptimist) [reply bazel/cc_proto_descriptor_library/file_descriptor_generator.cc]: Un-commented!

File: bazel/cc_proto_descriptor_library/repro/repro_test.cc (hunk 4 above)
Comment (RyanTheOptimist) [reply bazel/cc_proto_descriptor_library/repro/repro_test.cc]: Oh, actually I can just remove this whole directory. This is a set of protos I was using to help reproduce the problems I was having, but it's all obsolete now. Sorry, I should have caught this.

File: bazel/cc_proto_descriptor_library/repro/repro_test.cc (hunk 3 above)
Comment (RyanTheOptimist) [reply bazel/cc_proto_descriptor_library/repro/repro_test.cc]: Whoops. Removed.

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks!

**Review by abeyad — DISMISSED** [review]
Thanks Ryan!

**Review by mattklein123 — APPROVED** [review]
Just skimmed but this seems very cool and magical!

**Review by mattklein123 — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
> Just skimmed but this seems very cool and magical!

Thanks, Matt! The only wrinkle in this seems to be that [clang-tidy is complaining](https://dev.azure.com/cncf/envoy/_build/results?buildId=143633&view=logs&jobId=b7634614-24f3-5416-e791-4f3affaaed6c&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390). In particular, it's unable to find some files that are being `#include`'d. The code all builds fine, so clearly that's working. But clang-tidy can't find them. We've seen this a few times in other places in the codebase, I think, like with the logger. I suspect I'll need you to force submit this, unless you have any ideas about how to cajole clang-tidy into being happy?

**Comment by mattklein123** [comment]
Sorry no idea. We can force merge.

**Comment by phlax** [comment]
@RyanTheOptimist this failed clang-tidy before and after landing

- https://dev.azure.com/cncf/envoy/_build/results?buildId=143676&view=logs&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390&l=203
- https://dev.azure.com/cncf/envoy/_build/results?buildId=143675&view=logs&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390&l=205

the clang-tidy ci is not working optimally, not sure why, but it didnt create the diff just the fail yaml

either way would be good to clean up

cc @wbpcode

**Comment by RyanTheOptimist** [comment]
> @RyanTheOptimist this failed clang-tidy before and after landing
> 
> * https://dev.azure.com/cncf/envoy/_build/results?buildId=143676&view=logs&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390&l=203
> * https://dev.azure.com/cncf/envoy/_build/results?buildId=143675&view=logs&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390&l=205
> 
> the clang-tidy ci is not working optimally, not sure why, but it didnt create the diff just the fail yaml
> 
> either way would be good to clean up
> 
> cc @wbpcode

Thanks for noticing this. yeah, clang-tidy was really unhappy with this PR. Confusingly, the "formatting/style" was all fine. But the problem was that clang-tidy was unable to find various files which are #included. We've seen this in a couple of other places in Envoy. I really don't understand how clang is able to find the files when building but clang-tidy can't when tidying. Do you have any insight?

I also wondered if the was some -DENVOY_CLANG_TIDY we could pass in when tidy is running in order to #ifndef out code which tidy has irreconcilable problems with.

Of course actually fixing the issue would be way better.

**Pull request merged**